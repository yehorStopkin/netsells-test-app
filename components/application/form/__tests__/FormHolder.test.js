const { shallowMount } = require('@vue/test-utils');
const FormHolder = require('@/components/application/form/FormHolder').default;

describe('FormHolder', () => {
    getValue = () => {
        return {
            personalInfo: {
                firstName: 'Test first name',
                lastName: 'Test last name',
                email: 'Test email',
                phoneNumber: 'Test phone number',
            }, 
            additionalInfo: {
                liveInUk: true,
                gitProfile: 'Test git profile',
                aboutYou: 'Test about you',
            },
            additionalFiles: {
                cv: {},
                coverLetter: {},
            },
        };
    };

    getData = (filingOfPersonalInfo = 0, filingOfAdditionalInfo = 0, filingOfAdditionalFiles = 0, currentStep = 0, availableSteps = [0]) => {
        return {
            filingOfPersonalInfo,
            filingOfAdditionalInfo,
            filingOfAdditionalFiles,
            currentStep,
            availableSteps,
        };
    };

    test('component default', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData() } });
        expect(wrapper.exists()).toBeTruthy();
        expect(wrapper.element).toMatchSnapshot();

        const rootElement = wrapper.findAll('.form-holder');
        expect(rootElement.length).toBe(1);

        const formHolderHeader = wrapper.findAll('*[data-test-id="form-holder-header"]');
        expect(formHolderHeader.length).toBe(1);

        const dynamicComponent = wrapper.findAll('*[data-test-id="dynamic-component"]');
        expect(dynamicComponent.length).toBe(1);
    });

    test('check getter applicationForm', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData() } });

        expect(JSON.stringify(wrapper.vm.applicationForm)).toBe(JSON.stringify({
            personalInfo: wrapper.vm.innerPersonalInfo,
            additionalInfo: wrapper.vm.innerAdditionalInfo,
            additionalFiles: wrapper.vm.innerAdditionalFiles,
        }));
    });

    test('currentForm with PERSONAL_INFO', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData() } });

        expect(wrapper.vm.currentForm).toBe('form-personal-info');
    });

    test('currentForm with ADDITIONAL_INFO', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData(0, 0, 0, 1) } });

        expect(wrapper.vm.currentForm).toBe('form-additional-info');
    });

    test('currentForm with ADDITIONAL_INFO', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData(0, 0, 0, 2) } });

        expect(wrapper.vm.currentForm).toBe('form-additional-files');
    });

    test('check getter personalInfo', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData() } });

        expect(JSON.stringify(wrapper.vm.personalInfo)).toBe(JSON.stringify(wrapper.vm.innerPersonalInfo));
    });

    test('check getter additionalInfo', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData() } });

        expect(JSON.stringify(wrapper.vm.additionalInfo)).toBe(JSON.stringify(wrapper.vm.innerAdditionalInfo));
    });

    test('check getter additionalFiles', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData() } });

        expect(JSON.stringify(wrapper.vm.additionalFiles)).toBe(JSON.stringify(wrapper.vm.innerAdditionalFiles));
    });

    test('check onValidationChanged event', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData() } });

        expect(wrapper.vm.isPersonalInfoValid).toBe(false);
        wrapper.vm.onValidationChanged(0, true);
        expect(wrapper.vm.isPersonalInfoValid).toBe(true);

        expect(wrapper.vm.isAdditionalInfoValid).toBe(false);
        wrapper.vm.onValidationChanged(1, true);
        expect(wrapper.vm.isAdditionalInfoValid).toBe(true);

        expect(wrapper.vm.isAdditionalFilesValid).toBe(false);
        wrapper.vm.onValidationChanged(2, true);
        expect(wrapper.vm.isAdditionalFilesValid).toBe(true);
    });

    test('check stepSelected event from PERSONAL_INFO with invalid personal info', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData(0, 0, 0, 0) } });

        wrapper.vm.onStepSelected(0);
        expect(wrapper.emitted('step-selected')).not.toBeTruthy();

        wrapper.vm.onStepSelected(1);
        expect(wrapper.emitted('step-selected')).not.toBeTruthy();

        wrapper.vm.onStepSelected(2);
        expect(wrapper.emitted('step-selected')).not.toBeTruthy();
    });

    test('check stepSelected event from ADDITIONAL_INFO with invalid personal info', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData(0, 0, 0, 1) } });

        wrapper.vm.onStepSelected(0);
        expect(wrapper.emitted('step-selected')).not.toBeTruthy();

        wrapper.vm.onStepSelected(1);
        expect(wrapper.emitted('step-selected')).not.toBeTruthy();

        wrapper.vm.onStepSelected(2);
        expect(wrapper.emitted('step-selected')).not.toBeTruthy();
    });

    test('check stepSelected event from ADDITIONAL_FILES with invalid personal info', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData(0, 0, 0, 2) } });

        wrapper.vm.onStepSelected(0);
        expect(wrapper.emitted('step-selected')).not.toBeTruthy();

        wrapper.vm.onStepSelected(1);
        expect(wrapper.emitted('step-selected')).not.toBeTruthy();

        wrapper.vm.onStepSelected(2);
        expect(wrapper.emitted('step-selected')).not.toBeTruthy();
    });

    test('check stepSelected event from PERSONAL_INFO with valid personal info', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData(0, 0, 0, 0) } });

        wrapper.vm.isPersonalInfoValid = true;
        wrapper.vm.isAdditionalInfoValid = true;
        wrapper.vm.isAdditionalFilesValid = true;

        wrapper.vm.onStepSelected(0);
        expect(wrapper.emitted('step-selected')).toBeTruthy();

        wrapper.vm.onStepSelected(1);
        expect(wrapper.emitted('step-selected')).toBeTruthy();

        wrapper.vm.onStepSelected(2);
        expect(wrapper.emitted('step-selected')).toBeTruthy();
    });

    test('check stepSelected event from ADDITIONAL_INFO with valid personal info', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData(0, 0, 0, 1) } });

        wrapper.vm.isPersonalInfoValid = true;
        wrapper.vm.isAdditionalInfoValid = true;
        wrapper.vm.isAdditionalFilesValid = true;

        wrapper.vm.onStepSelected(0);
        expect(wrapper.emitted('step-selected')).toBeTruthy();

        wrapper.vm.onStepSelected(1);
        expect(wrapper.emitted('step-selected')).toBeTruthy();

        wrapper.vm.onStepSelected(2);
        expect(wrapper.emitted('step-selected')).toBeTruthy();
    });

    test('check stepSelected event from ADDITIONAL_FILES with valid personal info', () => {
        const wrapper = shallowMount(FormHolder, { propsData: { value: getValue(), ...getData(0, 0, 0, 2) } });

        wrapper.vm.isPersonalInfoValid = true;
        wrapper.vm.isAdditionalInfoValid = true;
        wrapper.vm.isAdditionalFilesValid = true;

        wrapper.vm.onStepSelected(0);
        expect(wrapper.emitted('step-selected')).toBeTruthy();

        wrapper.vm.onStepSelected(1);
        expect(wrapper.emitted('step-selected')).toBeTruthy();

        wrapper.vm.onStepSelected(2);
        expect(wrapper.emitted('step-selected')).toBeTruthy();
    });

});
