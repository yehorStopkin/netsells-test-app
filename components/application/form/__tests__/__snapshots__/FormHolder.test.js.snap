// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormHolder component default 1`] = `
<div
  class="form-holder"
>
  <form-holder-header-stub
    availablesteps="0"
    currentstep="0"
    data-test-id="form-holder-header"
    filingofadditionalfiles="0"
    filingofadditionalinfo="0"
    filingofpersonalinfo="0"
  />
   
  <form-personal-info-stub
    additional-files="[object Object]"
    additional-info="[object Object]"
    data-test-id="dynamic-component"
    personalinfo="[object Object]"
    step="0"
  />
</div>
`;
